### PR TITLE
[BEAM-1937] Free PTransform Names if they are being Replaced

### PR DESCRIPTION
**Pull request by tgroh**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [x] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---
Naming is based on what's in the graph, not what once was there.

### CODE REVIEW

**Comment by tgroh** [comment]
R: @dhalperi

**Comment by dhalperi** [comment]
Build doesn't pass.

**Comment by tgroh** [comment]
Should be fixed

**Comment by tgroh** [comment]
retest this please

**Comment by tgroh** [comment]
Build is successful, but got killed because it took too long to start

File: sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java [hunk 1]
```diff
@@ -230,25 +229,43 @@ public void visitPrimitiveTransform(Node node) {
...
+            if (clearingNamesFor != null) {
+              freedNames.add(node.getFullName());
+              return CompositeBehavior.ENTER_TRANSFORM;
+            }
             if (!node.isRootNode() && override.getMatcher().matches(node.toAppliedPTransform())) {
               matches.add(node);
               // This node will be replaced. It should not be visited.
-              return CompositeBehavior.DO_NOT_ENTER_TRANSFORM;
+              clearingNamesFor = node;
+              freedNames.add(node.getFullName());
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java]: It seems simpler and clearer to me to preserve the existing behavior (don't enter child).

Instead, add another function that is "collect all names to be replaces" that does an inner recursive traverse of the node to be replaced.

Traverses that do two things in modes based on local variables being null or not is ... confusing.

Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java]: Right now there's no way to traverse an arbitrary subgraph of a `TransformHierarchy` or `Pipeline`;

Instead I've changed this to track if the parent node is to-be-freed, starting with any replacement node. I think it's significantly more readable than the nullable variable, but still requires continuing to traverse the graph after we've matched a node to replace. Alternatively I could split this into two traversals. WDYT?

**Pull request closed without merging**